### ISSUE TITLE
no-value-for-parameter and control flow false positive.

### ISSUE THREAD

**Issue by RemiCardona** [PyCQA/pylint#2778]
### Steps to reproduce

Unfortunately, the following snippet does not trigger the error, but here's what our code roughly looks like

```python
class Abstract:
    def initialize(self, *args, **kwargs):
        raise NotImplementedError()

class A(Abstract):
    def initialize(self, a, b, c):
        pass

class B(Abstract):
    def initialize(self, d):
        pass

def main(iterable):
    for i in iterable:
        if i:
            obj = A()
            args = (1, 2, 3)
            break
    else:
        obj = B()
        args = (1,)

    obj.initialize(*args)  # <<< pylint complains here
```

### Current behavior
```
[E1120(no-value-for-parameter), main] No value for argument 'b' in method call
[E1120(no-value-for-parameter), main] No value for argument 'c' in method call
```

### Expected behavior

No error should be raised as `args` is set correctly along `obj` so they always match. Used to work just fine with pylint 2.2.2 and astroid 2.1.0 (and earlier versions).

### pylint --version output

```
pylint 2.3.0
astroid 2.2.0
Python 3.6.6 (default, Aug  9 2018, 11:28:52) 
[GCC 4.9.2]
```

How could I further debug the issue and get more info out of pylint?

**Comment by PCManticore** [comment]
@RemiCardona Thanks for reporting the issue.

It usually happens that with each release, `astroid` gets smarter as we're improving the inference engine continuously, so it's definitely possible that in one release for `pylint` to not emit a warning, but that warning will get emitted in a future release.

From the example you gave here (and as you said, it doesn't reproduce for this particular example), I would say it's a classic control flow issue, in which each branch can have a particular set of labels that go hand in hand together. For now `pylint` is definitely not smart enough to figure out that `args` will always conform to the `obj` that's set in the same branch. `pylint` is inferring all the possible branches together.

Right now the only solution is to disable the error in place, but hopefully at some point will have better inference for this kind of control flow.

**Comment by 0xLeon** [comment]
I'm not sure if this gives any new insights, but I just wanted to report that I'm able to trigger error 1120 with an even simpler construct, while also having a code example which is more or less exactly the same, but doesn't trigger this error.

### E1120 triggered
```Python
# pylint: disable=missing-docstring,too-few-public-methods,mixed-indentation,invalid-name

class Base():
	def __init__(self, foo, *args):
		self._foo = foo
		self._args = args

class Child(Base):
	pass


class SomethingElse():
	def __init__(self, *args):
		self._child = Child(*args)
```
#### Output
```
************* Module E1120-Trigger
E1120-Trigger.py:14:16: E1120: No value for argument 'foo' in constructor call (no-value-for-parameter)

-----------------------------------

Your code has been rated at 4.44/10
```

### E1120 not triggered
```Python
# pylint: disable=missing-docstring,too-few-public-methods,mixed-indentation,invalid-name

class Base():
	def __init__(self, foo, *args):
		self._foo = foo
		self._args = args

class Child(Base):
	def __init__(self, *args):
		super().__init__(*args)
		self._bar = True


class SomethingElse():
	def __init__(self, *args):
		self._child = Child(*args)
```
#### Output
```

------------------------------------

Your code has been rated at 10.00/10
```

### Conclusion
The difference is in the explicit call to the parent `__init__()` method in the second example. With this in place, E1120 is not triggered. In the first example, E1120 is triggered in the last line where a `Child` instance is created.

### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)]
```

**Comment by PCManticore** [comment]
@0xLeon Thanks for the additional example! Yours is something we can fix and would be better to have it separately from the original one.

**Comment by PCManticore** [comment]
@0xLeon I created https://github.com/PyCQA/pylint/issues/2820 to track the bug you mentioned.

**Comment by Pierre-Sassoulas** [comment]
Closing as we don't have a reproducer here, and we have a lot of other control flow issue that would most probably fix this if we handle them.
